Fix #683: Bundle all CSS/LESS resources from extensions

humphd opened the pull request:
This is coming along pretty well, but I still have some improvements to make.  My goal is to only as few separate CSS files as possible, by combining them all into a single minfied file.  I've modified the extension build steps to also include CSS/LESS files with what Brackets uses.

I have one load ordering bug, the `sidebarTheme.css` doesn't seem to be getting applied.  I wonder if it's the load order or something else.  I probably need some help from @flukeout or @Pomax debugging some time.  Otherwise, this is all working.

humphd commented:
BTW, testing this involves doing `npm run build` followed by browsing to `http://localhost:8000/dist/hosted.html`

humphd commented:
I was able to fix this.  However, I have a new problem.  Consider:

```
src/
  styles/
    brackets.less
  ...
  extensions/
    bramble/
      stylesheets/
        style.less
      images/
        folder-arrow.svg
```

Now we compile/combine all of the stylesheets into `dist/styles/brackets.min.css`:

```
dist/
  styles/
    brackets.min.css
  ...
  extensions/
    bramble/
      images/
        folder-arrow.svg
```

In the bramble extension's original stylesheet, we do the following:

```css
background-image: url(../images/folder-arrow.svg);
```

Which means that all relative paths now break because we've moved the stylesheet's location from `extension/bramble/stylesheets` to `style/`, and `../images` can't be found.

What is the right fix for this?  I'm trying to decide if I should just throw this away (which sucks, because it reduces the number of requests we make), or just have each extension write its own compiled css file (still good, because we can switch back to less everywhere).

@flukeout, @gideonthomas, @Pomax Help?

gideonthomas commented:
@humphd not sure which one is easier but we can have each extension generate its own compiled css file as you mention OR, if we are seeing a pattern (specifically for images), can't we do the same for images and just dump all of them in `dist/images/...`? Will that work?

humphd commented:
The problem with trying to follow a pattern (e.g., copy all images into `dist/images`) is that every extension does it differently, and we'll be back to having to remember to "fix" any that don't follow our pattern.

Maybe I should just write each CSS file our per-extension and then it will always work with whatever they do.  Just sucks, because it would be nice to not have all these unnecessary requests.

flukeout commented:
Could we rewrite our css path references to always go from the root instead of being relative or something like that?

flukeout commented:
Would you be able to rewrite the URL paths when you compile? So you would end up converting relative to absolute paths as you go?

humphd commented:
I don't think I can do it without making some elaborate, custom solution, which is likely going to break on some edge case I can't anticipate.  I wondered if the `grunt-contrib-less` stuff could do it, but all I found was https://github.com/gruntjs/grunt-contrib-less#relativeurls.

flukeout commented:
Can we bundle all extension images into a single folder and then reference that location? We'd probably get filename collisions though huh.

humphd commented:
Filename collisions, plus, there is no standard that says an extension looks for things in the same path (same for all assets).  Every extension is the the wild west.

I think I have to bundle-per-extension.  But, at least I can add build-time LESS->CSS compiling and minifiying, so it's a partial win.

I'll fix this PR to do that, and then it's probably ready for review.

humphd commented:
@gideonthomas, @flukeout this is ready for review.  I've done the following:

* switched all our `.css` to `.less` for easier development.  We should always use `.less` going forward in extensions and Brackets in general.
* I've removed a bunch of `.css` files in extensions that I had manually created from their associated `.less` files.
* I've reverted all the changes I made to extensions in the past to load `.css` vs. `.less`.  We'll only ever do this in `src/ `builds.
* I've combined the `codemirror.css` into our `brackets.min.css` so we don't have to load an extra file
* I've combined our `sidebarTheme.less` with our bramble `styles.css` so we don't have to load 2 files.
* I've updated our extension build stuff to add proper less support for every extension that needs it.  We'll generate `.css` for all those files listed
* I've hacked our `.less` and `.css` loading to always use `.css` in production builds

I think everything works fine, but try doing a `src/` build browsing `localhost:8000/src/hosted.html` and also `npm run build` and then `localhost:8000/dist/hosted.html`.  See if I missed anything.

gideonthomas commented on Gruntfile.js:
```diff
@@ -435,6 +444,13 @@ module.exports = function (grunt) {
         });
 
+        // Copy any LESS/CSS files from extensions to the less task file list.
+        extensions.forEach(function(extension) {
```
I think as a follow-up, one optimization we could do is to rely on `fs.find` to list all the files with `.less` for each extension and generate the dist css file based on that path instead of manually storing it in the `bramble-extensions.json` file.
